### ISSUE TITLE
How to create a new layer of Kur

### ISSUE THREAD

**Issue by EmbraceLife** [EmbraceLife/LIE#2]
Concerning how to create a new layer directly using tensorflow backend: 

Q1: given the current kur, can we write a new layer directly use tensorflow as backend, instead of using keras as backend? 
Q2: my observation of `kur/containers/layers/` suggests that we can only build a new layer either using `keras` backend or `pytorch` backend,  kur seems not using  `theano` or `tensorflow` backend directly to build a layer, is this observation correct?
Q3: Suppose we see a new layer invented by a user using tensorflow (not yet be tried with keras), if we can't directly in kur to write this layer using tensorflow backend, then does it mean keras can easily write most if not all layers created with tensorflow, so we don't need bother to write with tensorflow backend?

@ajsyp response: 
> Kur's Keras backend is just that: a backend which uses Keras. Keras will transparently use either Theano or TensorFlow as available. So yes, Kur isn't using direct Theano/TensorFlow code (otherwise we'd have to maintain 2 backends rather than one)--much easier to build on the existing, excellent work by the Keras team.
> That said, yes, you can use Theano/TensorFlow code within Kur's Keras backend. Obviously, if you add Theano-only code (for example), then it may not run on TensorFlow, so if you need some particular functionality from Theano/TF, you should either fail gracefully if the user requests an unsupported backend for your layer, or you should write both TH/TF versions.
> There is no reason that separate Theano-/TensorFlow-only backends couldn't be written.
> I'm not sure I understand your "Q3". If pure-TensorFlow layer is written, it needs to work with a Kur backend. So if a pure-TensorFlow backend is also written, then no problem. Otherwise, you'll need to reuse an existing backend. What if that existing backend were the Keras backend? Well, great, but you still need to shove the TensorFlow layers into the Keras backend, using the Keras-related API.
And of course, it might not work if Keras loads the Theano backend.
A perfect example of this is the multi-GPU support in Kur.
It isn't a layer, but it uses TensorFlow-specific code, merged into Keras, to support multi-GPU. Theano doesn't have equivalent functionality, and so multi-GPU doesn't work with Theano.

**Comment by EmbraceLife** [comment]
How exactly to write a layer (which is ready in keras) in kur using keras backend: 
- Compare layer in keras and layers in kur built with keras side by side

Basic Idea: 
- prepare args 
- import keras.backend and keras.layers for layers 
- instantiate those layers with args defined above, and yield 

Workflow: 
- `__init__()`: super class and initialize some attrs or args inputs
- `_parse()`: set the attrs or args to specific values according to kurfile 
- `_build()`: wrap keras' layer and instantiate it for use in kur 
- `shape()`: access the shape of the current layer (not used in `build()` or `train()`)
- Note: `placeholder` layer has more methods than the other layers

**Comment by EmbraceLife** [comment]
**Layer**: see [source](https://github.com/EmbraceLife/LIE/blob/master/kur/containers/layers/layer.py)
- is a subclass of [Container](https://github.com/EmbraceLife/LIE/blob/master/kur/containers/container.py) class: has lots of attributes and methods
- Base class for layers, which are containers that produce backend-specific layers, 
- or which wrap backend-specific layers with additional backend-specific operations.
- it is almost empty

**Comment by EmbraceLife** [comment]
How to [write your own layer in keras](https://github.com/fchollet/keras/blob/7f58b6fbe702c1936e88a878002ee6e9c469bc77/docs/templates/layers/writing-your-own-keras-layers.md)?
- dive into later

**Comment by EmbraceLife** [comment]
`**kwargs` as arg inputs to `L.Conv2D`: 

- `trainable = True or False` is an example of `**kwargs` for `L.Conv2D`
- `L.Input` has no args like `trainable`, because this layer needs not training, and plus this layer `__init__` has no `**kwargs`
- `L.Dense` has `trainable` args, as its weights need to be trained, also its `__init__` has `**kwargs`
- `L.Dropout` has no `trainable` args, cos there is no weights to be trained here, although its `__init__` has `**kwargs`

**Comment by EmbraceLife** [comment]
I have added `LeakyReLU` from `keras`: 
- see [commit](https://github.com/EmbraceLife/LIE/commit/01551aa0f558b5be4c2c9f38e7ca3d55de2da443)
- commit above explains how to use `LeakyReLU` and other activations inside kurfile in detail

Note: 
- we can't write: 
```yaml
activation: relu
```
- we must write 
```yaml
activation: 
  name: relu
```
